src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,37 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
 
 public class AsyncRequestProcessor {
+    private final Map<String, UserData> cache = new ConcurrentHashMap<>();
+    private final Function<String, UserData> createUserDetails;
     private final Executor executor;
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
+        createUserDetails = userId -> new UserData(userId, "Details for " + userId);
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        return CompletableFuture.supplyAsync(() -> fetchUserData(userId), executor)
+                .exceptionally(throwable -> {
+                    System.out.println(throwable);
+                    return null;
+                });
+    }
+
+    private UserData fetchUserData(String userId) {
+        try {
+            TimeUnit.MILLISECONDS.sleep(300);
+
+            return cache.computeIfAbsent(userId, createUserDetails);
```

Comment:
this is wrong actually

you should fetch data first, and only after that store it in cache


in other words you should:
```
if (!cache.contains(userId)) {
return CompletableFuture.supplyAsync(() -> userDataFunction.apply(userId)) // no cache involved here
.whenComplete((userData, ex) -> {
    if (ex != null) {
       throw ex;
    }
    cache.put(userId, userData);
   })
}
// retrieve from cache otherwise
```